File: 백준/Gold/1717. 집합의 표현/집합의 표현.cc
#include <iostream>
#include <cstdio>

using namespace std;

#define MAXX 1000001 

int parent[MAXX], n, m, check, first, second;

void _union(int, int); int find(int); int main() {
	ios_base::sync_with_stdio(false); cin.tie(0); cin >> n >> m;

	for (int i = 1; i <= n; i++) { parent[i] = i; } while (m--) {
		cin >> check >> first >> second;
		if (!check) { _union(first, second); }
		else {
			if (find(first) == find(second)) {
				printf("YES\n");
			}
			else {
				printf("NO\n");
			}
		}
	}
}
void _union(int f, int s) {
	int firstParent = find(f);
	int secondParent = find(s);

	if (firstParent != secondParent)
		parent[firstParent] = secondParent;

} int find(int i) {
	if (parent[i] == i) {
		return i;
	}
	else {
		return parent[i] = find(parent[i]);
	}
}

File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
import sys

sys.setrecursionlimit(10 ** 5)
def input(): return sys.stdin.readline().rstrip()


n, m = map(int, input().split())
parent = [i for i in range(n + 1)]


def find(x):
    if x == parent[x]:
        return x

    parent[x] = find(parent[x])

    return parent[x]


def union(a, b):
    a = find(parent[a])
    b = find(parent[b])

    if a > b:
        parent[a] = b
    else:
        parent[b] = a


for _ in range(m):
    c, a, b = map(int, input().split())
    if c == 0:
        union(a, b)
    elif c == 1:
        if find(parent[a]) == find(parent[b]):
            print('YES')
        else:
            print('NO')